File: node/ts/test/chat/UnauthLoginPurchaseServiceTest.ts
//
// Copyright 2026 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

import { config, expect, use } from 'chai';
import chaiAsPromised from 'chai-as-promised';

import * as NativeNice from '../../NativeNice.js';
import * as util from '../util.js';
import { UnauthLoginPurchaseService } from '../../net.js';
import { connectUnauth, defineTestGrpcCases } from './ServiceTestUtils.js';
import { ErrorCode, LibSignalErrorBase } from '../../Errors.js';
import { ServerPublicParams } from '../../zkgroup/index.js';

use(chaiAsPromised);

util.initLogger();
config.truncateThreshold = 0;

describe('UnauthLoginPurchaseService', () => {
  describe('createLoginReceiptCredential', () => {
    defineTestGrpcCases(
      NativeNice.TESTING_CreateLoginReceiptCredentialTests(),
      connectUnauth<UnauthLoginPurchaseService>,
      async (
        chat: UnauthLoginPurchaseService,
        {
          paymentProcessor,
          purchaseIdentifier,
          receiptCredentialRequestContext,
          serverParams,
          purchaseTime,
        }: NativeNice.CreateLoginReceiptCredentialArgs,
        resp: NativeNice.CreateLoginReceiptCredentialOut
      ) => {
        const out = chat.createLoginReceiptCredential({
          paymentProcessor,
          purchaseIdentifier,
          receiptCredentialRequestContext,
          serverParams: new ServerPublicParams(serverParams.bytes),
          purchaseTime,
        });
        if ('success' in resp) {
          expect(await out).to.deep.equal(resp.success);
        } else if ('unexpectedError' in resp) {
          await expect(out).to.eventually.be.rejectedWith(resp.unexpectedError);
        } else if ('explicitError' in resp) {
          if (typeof resp.explicitError === 'string') {
            const codes: Record<
              | 'paymentNotFound'
              | 'paymentStillProcessing'
              | 'receiptAlreadyIssued',
              ErrorCode
            > = {
              paymentNotFound: ErrorCode.ReceiptCredentialErrorPaymentNotFound,
              paymentStillProcessing:
                ErrorCode.ReceiptCredentialErrorPaymentStillProcessing,
              receiptAlreadyIssued:
                ErrorCode.ReceiptCredentialErrorReceiptAlreadyIssued,
            };
            const code = codes[resp.explicitError];
            if (code === undefined) {
              throw new Error('Invalid code!');
            }
            await expect(out)
              .to.eventually.be.rejectedWith(LibSignalErrorBase)
              .and.deep.include({ code });
          } else {
            const chargeFailure =
              resp.explicitError.paymentRequired.length > 0
                ? resp.explicitError.paymentRequired[0]
                : null;
            await expect(out)
              .to.eventually.be.rejectedWith(LibSignalErrorBase)
              .and.deep.include({
                code: ErrorCode.ReceiptCredentialErrorPaymentRequired,
                chargeFailure,
              });
          }
        } else {
          resp satisfies never;
          throw new Error('Unreachable!');
        }
      }
    );
  });
});

File: java/client/src/test/java/org/signal/libsignal/net/UnauthLoginPurchaseServiceTest.kt
//
// Copyright 2026 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

package org.signal.libsignal.net

import kotlinx.coroutines.test.runTest
import org.signal.libsignal.internal.CreateLoginReceiptCredentialOut
import org.signal.libsignal.internal.NativeTestingNice
import org.signal.libsignal.internal.ReceiptCredentialError
import org.signal.libsignal.net.assertNonSuccess
import org.signal.libsignal.zkgroup.ServerPublicParams
import org.signal.libsignal.zkgroup.receipts.ReceiptCredential
import kotlin.test.Test
import kotlin.test.assertContains
import kotlin.test.assertEquals
import kotlin.test.assertIs

class UnauthLoginPurchaseServiceTest {
  @Test
  fun testCreateLoginReceiptCredential() {
    runTest {
      GrpcTestCase.runTests(
        NativeTestingNice.TESTING_CreateLoginReceiptCredentialTests(),
        { tokio, listener ->
          UnauthenticatedChatConnection.fakeConnect(tokio, listener, Network.Environment.STAGING)
        },
        ::UnauthLoginPurchaseService,
        invoke = { chat, req ->
          chat.createLoginReceiptCredential(
            paymentProcessor = req.paymentProcessor,
            purchaseIdentifier = req.purchaseIdentifier,
            receiptCredentialRequestContext = req.receiptCredentialRequestContext,
            serverParams = ServerPublicParams(req.serverParams.bytes),
            purchaseTime = req.purchaseTime,
          )
        },
        check = { expected, actual ->
          when (expected) {
            is CreateLoginReceiptCredentialOut.ExplicitError ->
              when (expected._0) {
                ReceiptCredentialError.PaymentNotFound ->
                  actual
                    .assertNonSuccess<_, _, CreateLoginReceiptCredentialException.PaymentNotFound>()
                is ReceiptCredentialError.PaymentRequired ->
                  assertEquals(
                    expected._0.chargeFailure.firstOrNull(),
                    actual
                      .assertNonSuccess<_, _, CreateLoginReceiptCredentialException.PaymentRequired>()
                      .chargeFailure,
                  )
                ReceiptCredentialError.PaymentStillProcessing ->
                  actual
                    .assertNonSuccess<_, _, CreateLoginReceiptCredentialException.PaymentStillProcessing>()
                ReceiptCredentialError.ReceiptAlreadyIssued ->
                  actual
                    .assertNonSuccess<_, _, CreateLoginReceiptCredentialException.ReceiptAlreadyIssued>()
              }
            is CreateLoginReceiptCredentialOut.Success ->
              assertEquals(
                expected._0,
                assertIs<RequestResult.Success<ReceiptCredential>>(actual).result,
              )
            is CreateLoginReceiptCredentialOut.UnexpectedError ->
              assertContains(assertIs<RequestResult.ApplicationError>(actual).toString(), expected.contains)
          }
        },
      )
    }
  }
}
